cpp/source/deal.II/geometry.templates.h

```diff
@@ -226,83 +216,71 @@ Geometry<dim>::Geometry(std::shared_ptr<boost::property_tree::ptree> database,
-void Geometry<dim>::repartition(
-    std::shared_ptr<boost::property_tree::ptree const> database)
-{
-  std::array<unsigned int, 4> weights;
-  weights[internal::WEIGHT_TYPE::anode] = database->get("anode.weight", 0);
-  weights[internal::WEIGHT_TYPE::cathode] = database->get("cathode.weight", 0);
-  weights[internal::WEIGHT_TYPE::separator] =
-      database->get("separator.weight", 0);
-  weights[internal::WEIGHT_TYPE::collector] =
-      database->get("collector.weight", 0);
-  _triangulation->signals.cell_weight.connect(
-      std::bind(&Geometry<dim>::compute_cell_weight, this,
-                std::placeholders::_1, weights));
-  _triangulation->repartition();
-}
-
-template <int dim>
-unsigned int Geometry<dim>::compute_cell_weight(
-    typename dealii::Triangulation<dim, dim>::cell_iterator const &cell,
-    std::array<unsigned int, 4> const &weights) const
+void Geometry<dim>::repartition()
 {
   // Cells in the anode of the cathode have to deal with two physics instead of
   // only one in the collectors and the separator. Each cell starts with a
   // default weight of 1000. This function returns the extra weight on some of
   // the cells.
-  dealii::types::material_id material = cell->material_id();
-  std::vector<dealii::types::material_id> &anode = (*_materials)["anode"];
-  std::vector<dealii::types::material_id> &cathode = (*_materials)["cathode"];
-  std::vector<dealii::types::material_id> &separator =
-      (*_materials)["separator"];
-  std::vector<dealii::types::material_id> &collector =
-      (*_materials)["collector"];
-  if (std::find(anode.begin(), anode.end(), material) != anode.end())
-    return weights[internal::WEIGHT_TYPE::anode];
-  if (std::find(cathode.begin(), cathode.end(), material) != cathode.end())
-    return weights[internal::WEIGHT_TYPE::cathode];
-  if (std::find(separator.begin(), separator.end(), material) !=
-      separator.end())
-    return weights[internal::WEIGHT_TYPE::separator];
-  if (std::find(collector.begin(), collector.end(), material) !=
-      collector.end())
-    return weights[internal::WEIGHT_TYPE::collector];
-  return dealii::numbers::invalid_unsigned_int;
+  auto compute_cell_weight = [=](
+      typename dealii::Triangulation<dim, dim>::cell_iterator const &cell,
+      typename dealii::Triangulation<dim, dim>::CellStatus const) mutable
+      -> unsigned int
+  {
+    for (auto const &m : *_materials)
+    {
+      if (m.second.count(cell->material_id()) > 0)
+        return _weights[m.first];
+    }
+    throw std::runtime_error("Cell material id is not listed.");
+    return dealii::numbers::invalid_unsigned_int; // @Bruno: should we throw
+                                                  // or return invalid weight?
+                                                  // It does not seem to throw
+                                                  // (in release mode at least)
+                                                  // and this does unnoticed.
```

Comment:
It should throw because I don't think that deal.II checks the value you give. I don't know why it doesn't work.